### ISSUE TITLE
UnicodeEncodeError at /projects/test2/versions/

### ISSUE THREAD

**Issue by NikitaEgorov** [rtfd/readthedocs.org#2506]
## Details

* Project URL: local
* Build URL (if applicable): local
* Read the Docs username (if applicable): local

## Actual Result
UnicodeEncodeError at /projects/test2/versions/
'ascii' codec can't encode characters in position 19-26: ordinal not in range(128)

```
Request Method:	GET
Request URL:	http://****:8000/projects/test2/versions/
Django Version:	1.8.3
Exception Type:	UnicodeEncodeError
Exception Value:	
'ascii' codec can't encode characters in position 19-26: ordinal not in range(128)
Exception Location:	/home/tempuser/checkouts/readthedocs.org/readthedocs/builds/models.py in identifier_friendly, line 178
Python Executable:	/usr/bin/python
Python Version:	2.7.6
Python Path:	
['/home/tempuser/checkouts/readthedocs.org',
 '/usr/lib/python2.7',
 '/usr/lib/python2.7/plat-x86_64-linux-gnu',
 '/usr/lib/python2.7/lib-tk',
 '/usr/lib/python2.7/lib-old',
 '/usr/lib/python2.7/lib-dynload',
 '/usr/local/lib/python2.7/dist-packages',
 '/usr/lib/python2.7/dist-packages',
 '/usr/lib/python2.7/dist-packages/PILcompat',
 '/usr/lib/python2.7/dist-packages/gtk-2.0',
 '/usr/lib/python2.7/dist-packages/ubuntu-sso-client',
 '/home/tempuser/checkouts/readthedocs.org']
Server time:	Tue, 15 Nov 2016 06:45:50 -0600
```

Unicode error hint

The string that could not be encoded/decoded was: -941_Добавить_на_ф
Error during template rendering

In template /home/tempuser/checkouts/readthedocs.org/readthedocs/templates/projects/project_version_list.html, error at line 102

```

92	                  <span class="quiet">{{ version.slug }}</span>
93	
94	
95	                  {% if request.user|is_admin:project %}
96	                    <span class="right-menu">
97	                      {{ version.get_privacy_level_display }}
98	                    </span>
99	                  {% endif %}
100	
101	                  {% if not version.slug in version.identifier_friendly %}
102	
                          <span class="right-menu quiet version-branch">
      {{ version.identifier_friendly|truncatechars:24 }}
      </span>

      
103	                  {% endif %}
104	
105	                  {% if request.user|is_admin:project %}
106	                    <ul class="module-item-menu">
107	                      {% if version.built %}
108	                          <li><a href="{% url "project_version_delete_html" project.slug version.slug %}">{% trans "Clean" %}</a></li>
109	                      {% endif %}
110	                      <li><a href="{% url "wipe_version" project.slug version.slug %}">{% trans "Wipe" %}</a></li>
111	                      <li><a href="{% url "project_version_detail" project.slug version.slug %}">{% trans "Edit" %}</a></li>
112	                    </ul>
```

**Comment by agjohnson** [comment]
Thanks for the report! Looks like a bug with the `truncatechars` perhaps? This would be an easy first contribution to RTD if anyone wants to hunt this down.

**Comment by PeterJCLaw** [comment]
Interesting. I can get a similar error though from the `sort_version_aware` on line 87.
My steps to reproduce:
- create local test repo with some docs
- create a tag in that repo (I used `1.0.0`)
- import that into a local RTD instance, build its docs
- visit http://127.0.0.1:8000/projects/test/versions/, to validate it works
- create a new tag in the repo which contains the "bad" characters (I used a tag called `941_Добавить_на_ф`)
- update the local instance (`./manage.py update_repos test`)
- visit http://127.0.0.1:8000/projects/test/versions/, to see error.

The trace I get is:
```
File "/home/peter/.virtualenvs/rtd/local/lib/python2.7/site-packages/django/template/base.py" in resolve
  675.                 new_obj = func(obj, *arg_vals)
File "/home/play/readthedocs.org/readthedocs/projects/templatetags/projects_tags.py" in sort_version_aware
  17.         reverse=True)
File "/home/play/readthedocs.org/readthedocs/projects/templatetags/projects_tags.py" in <lambda>
  16.         key=lambda version: comparable_version(version.verbose_name),
File "/home/play/readthedocs.org/readthedocs/projects/version_handling.py" in comparable_version
  122.     comparable = parse_version_failsafe(version_string)
File "/home/play/readthedocs.org/readthedocs/projects/version_handling.py" in parse_version_failsafe
  110.         return Version(version_string)
File "/home/peter/.virtualenvs/rtd/local/lib/python2.7/site-packages/packaging/version.py" in __init__
  212.             raise InvalidVersion("Invalid version: '{0}'".format(version))

Exception Type: UnicodeEncodeError at /projects/test/versions/
Exception Value: 'ascii' codec can't encode characters in position 4-11: ordinal not in range(128)
```
Which suggests that this is a secondary issue to the tag not being parsed cleanly. I'm not entirely sure this is the same issue, but I'll dig into it some more tomorrow.

**Comment by PeterJCLaw** [comment]
I don't think that `truncatechars` is to blame here. In part as I can't get that to emit an `UnicodeEncodeError` using the sample string, but also since the original report includes:
```
Exception Location: <snip>/readthedocs/builds/models.py in identifier_friendly, line 178
```
which I'd missed earlier.

(I also think the other issue I found is unrelated, though I've reported that upstream).

It looks like we can hit a `UnicodeEncodeError` in `Version.identifier_friendly` when the `Version.identifier` contains unicode (according to the inline docs, that field can contain branch names for some VCSes). The [code in question](https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/builds/models.py#L175) calls `str` on the `identifier` field before trying to match it to a regex. Unfortunately the history doesn't tell us _why_ it does that.

@agjohnson It looks like you authored that code, do you happen to know why the call to `str` is there?

Given that the `identifier` field will always contain a `unicode` value (on Python 2) and that the `re` library seems quite happy to be given a `unicode` instance, the fix seems to be to remove the `str` call from that line.

**Comment by PeterJCLaw** [comment]
For reference, the steps I'm using to reproduce the original issue are:
- create a test project
- create a branch in that test project which contains non-ascii (for me `git branch _Добавить_на_ф`)
- import said project
- view the `/projects/$proj/versions/` page

This is fixed by removing the `str()` call, and all the existing tests pass so I've bundled the fix (and a test) into #2560.
